File: mobile/services/shareSave.ts
import * as Notifications from 'expo-notifications';
import { Platform } from 'react-native';
import { api } from './api';
import { fetchClientMetadata } from './clientExtract';
import { refreshUsage } from './usageCache';

/**
 * Save a shared link WITHOUT the user having to look at SaveHere.
 *
 * ⚠️ THE POINT OF THIS FILE. A share is an interruption of something else —
 * you are mid-scroll in Instagram, you want the reel kept, you want to carry on
 * scrolling. Routing that to /save and making the user watch a progress screen
 * breaks exactly the flow the feature exists to protect. Nothing here needs a
 * decision from them: the URL is known, the save is unconditional, and the
 * summary was always asynchronous anyway.
 *
 * ⚠️ PHASE A. Android still LAUNCHES the app to deliver ACTION_SEND, so there
 * is a brief flash of SaveHere before it hands control back. Removing that
 * needs a translucent no-display Activity (a custom config plugin + Kotlin),
 * which is Phase B. iOS needs a Share Extension plus an App Group to share the
 * session — blocked on the Apple Developer account. This module is written so
 * both of those become "call `saveSharedLink` from somewhere else" rather than
 * a rewrite.
 */

/** Human name for the notification, from the URL alone — the server's platform
 *  field is not back yet when we post it, and waiting for it would defeat the
 *  purpose. */
export function platformLabel(url: string): string {
  const u = url.toLowerCase();
  if (u.includes('youtube.com') || u.includes('youtu.be')) return 'YouTube';
  if (u.includes('instagram.com')) return 'Instagram';
  if (u.includes('facebook.com') || u.includes('fb.watch') || u.includes('fb.com')) return 'Facebook';
  if (u.includes('tiktok.com')) return 'TikTok';
  if (u.includes('linkedin.com')) return 'LinkedIn';
  return 'the web';
}

/**
 * Ask once, and never block the save on the answer.
 *
 * ⚠️ Android 13+ requires runtime POST_NOTIFICATIONS. A user who declines still
 * gets their save — the notification is how we REPORT the work, not the work
 * itself, so a refusal must never cost them the link.
 */
async function canNotify(): Promise<boolean> {
  try {
    const { status } = await Notifications.getPermissionsAsync();
    if (status === 'granted') return true;
    const asked = await Notifications.requestPermissionsAsync();
    return asked.status === 'granted';
  } catch {
    return false;
  }
}

async function notify(title: string, body: string) {
  if (!(await canNotify())) return;
  try {
    await Notifications.scheduleNotificationAsync({
      content: { title, body },
      trigger: null,   // immediately
    });
  } catch {
    // A failed notification must never surface as a failed save.
  }
}

/**
 * Android needs a channel or notifications are silently dropped on 8.0+.
 * Idempotent, so calling it on every share is fine.
 */
export async function ensureNotificationChannel() {
  if (Platform.OS !== 'android') return;
  try {
    await Notifications.setNotificationChannelAsync('saves', {
      name: 'Saved links',
      importance: Notifications.AndroidImportance.DEFAULT,
    });
  } catch {}
}

/**
 * The whole background save. Resolves when the card exists server-side; the AI
 * summary continues on the backend after that, exactly as it does for a normal
 * save.
 *
 * Returns whether it worked, so the caller can decide whether to bounce the
 * user back out or leave the app open on an error they should see.
 */
export async function saveSharedLink(url: string): Promise<boolean> {
  const where = platformLabel(url);
  await ensureNotificationChannel();
  try {
    // Same two-step as app/save.tsx: kick the device-side metadata fetch off
    // alongside the save (never in front of it), then deliver it after. On
    // Instagram and Facebook this is the ONLY path that gets a caption, because
    // the server's datacenter IP is served a login wall.
    const metaPromise = fetchClientMetadata(url);
    const reel = await api.saveReel(url);
    metaPromise
      .then(meta => (meta ? api.sendClientMetadata(reel.id, meta) : null))
      .catch(() => {});

    // The AI budget just moved; keep the cache honest for the next screen.
    refreshUsage();

    await notify(
      `Saved from ${where}`,
      'Your summary is being written — it will be ready in your library.',
    );
    return true;
  } catch (e: any) {
    // Say what happened. A silent failure here is the worst outcome of all:
    // the user believes the link is kept, carries on scrolling, and finds
    // nothing later.
    await notify(
      `Couldn't save that ${where} link`,
      e?.message || 'Open SaveHere and paste it to try again.',
    );
    return false;
  }
}

File: mobile/plugins/android/ShareSave.kt
package com.savehere.app.share

import android.app.Activity
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.database.sqlite.SQLiteDatabase
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import androidx.core.app.NotificationCompat
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

/**
 * ANDROID PHASE B — the invisible share.
 *
 * Phase A launched the whole app to receive ACTION_SEND: it worked, but it
 * flashed SaveHere for 1–2 s in the middle of someone's Instagram scroll, which
 * is the exact interruption the share feature exists to avoid. Everything here
 * exists to delete that flash.
 *
 * ⚠️ THIS RUNS OUTSIDE THE JS RUNTIME. There is no React Native context and no
 * supabase-js, so it cannot use the Supabase session:
 *   * a stored access token lives ~1 h and only refreshes while the app is
 *     open, so most shares would 401;
 *   * refreshing from here would revoke the refresh token the app still holds
 *     and sign the user out on the next launch.
 * It carries a save-scoped share key instead — see backend/app/sharekey.py and
 * mobile/services/shareKey.ts.
 *
 * ⚠️ IT ALSO CANNOT CALL saveSharedLink(). Everything the JS path does that
 * still matters here is reproduced: the POST, the notification, and appending
 * to the notification drawer's AsyncStorage row. What is deliberately NOT
 * reproduced is the client-side metadata fetch — as of 2026-08-13 the server
 * reads Instagram captions itself through the embed route and Facebook through
 * oEmbed (docs/EXTRACTION_ROUTES.md), so the phone no longer has to lend its
 * residential IP for the platforms that mattered.
 */

/* ── Storage: AsyncStorage is SQLite on Android, not SharedPreferences ─────── */

private const val SHARE_KEY_ROW = "@savehere:sharekey:v1"
private const val NOTES_ROW = "@savehere:notifications:v1"
private const val MAX_NOTES = 10
private const val CHANNEL_ID = "saves"

internal data class ShareConfig(val apiUrl: String, val key: String)

/**
 * Read one AsyncStorage row.
 *
 * `RKStorage` (table `catalystLocalStorage`) is the file @react-native-async-storage
 * writes. Read-only, and every failure returns null so a schema change in a
 * future AsyncStorage release degrades to "open the app" rather than crashing
 * in someone's share sheet.
 */
internal fun readRow(ctx: Context, key: String): String? {
    val file = ctx.getDatabasePath("RKStorage")
    if (file == null || !file.exists()) return null
    var db: SQLiteDatabase? = null
    return try {
        db = SQLiteDatabase.openDatabase(file.path, null, SQLiteDatabase.OPEN_READONLY)
        db.rawQuery("SELECT value FROM catalystLocalStorage WHERE key = ?", arrayOf(key)).use { c ->
            if (c.moveToFirst()) c.getString(0) else null
        }
    } catch (t: Throwable) {
        null
    } finally {
        try { db?.close() } catch (t: Throwable) { }
    }
}

internal fun writeRow(ctx: Context, key: String, value: String) {
    val file = ctx.getDatabasePath("RKStorage")
    if (file == null || !file.exists()) return
    var db: SQLiteDatabase? = null
    try {
        db = SQLiteDatabase.openDatabase(file.path, null, SQLiteDatabase.OPEN_READWRITE)
        db.execSQL(
            "INSERT OR REPLACE INTO catalystLocalStorage (key, value) VALUES (?, ?)",
            arrayOf<Any>(key, value)
        )
    } catch (t: Throwable) {
        // The app may hold a write lock. Losing a drawer entry is survivable;
        // crashing the share is not.
    } finally {
        try { db?.close() } catch (t: Throwable) { }
    }
}

internal fun readConfig(ctx: Context): ShareConfig? {
    val raw = readRow(ctx, SHARE_KEY_ROW) ?: return null
    return try {
        val o = JSONObject(raw)
        val url = o.optString("apiUrl").trimEnd('/')
        val key = o.optString("key")
        if (url.isEmpty() || key.isEmpty()) null else ShareConfig(url, key)
    } catch (t: Throwable) {
        null
    }
}

/**
 * Append to the notification drawer the profile panel reads.
 *
 * ⚠️ Without this, the ONLY notifications missing from the in-app list would be
 * the ones from invisible shares — i.e. exactly the saves the user never saw
 * happen. Shape and cap must match mobile/services/notifyLog.ts.
 */
internal fun appendNote(ctx: Context, title: String, body: String) {
    try {
        val now = System.currentTimeMillis()
        val note = JSONObject()
            .put("id", now.toString() + "-" + UUID.randomUUID().toString().take(6))
            .put("title", title)
            .put("body", body)
            .put("at", now)
            .put("read", false)
        val existing = try { JSONArray(readRow(ctx, NOTES_ROW) ?: "[]") } catch (t: Throwable) { JSONArray() }
        val out = JSONArray().put(note)
        var i = 0
        while (i < existing.length() && out.length() < MAX_NOTES) {
            existing.optJSONObject(i)?.let { out.put(it) }
            i++
        }
        writeRow(ctx, NOTES_ROW, out.toString())
    } catch (t: Throwable) {
    }
}

/* ── Notifications ────────────────────────────────────────────────────────── */

internal fun ensureChannel(ctx: Context) {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
    val mgr = ctx.getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager ?: return
    // Same id the JS path uses (services/shareSave.ts), so both land in one
    // channel and the user has one switch, not two.
    mgr.createNotificationChannel(
        NotificationChannel(CHANNEL_ID, "Saved links", NotificationManager.IMPORTANCE_DEFAULT)
    )
}

internal fun buildNotification(ctx: Context, title: String, body: String, ongoing: Boolean): android.app.Notification {
    val launch = ctx.packageManager.getLaunchIntentForPackage(ctx.packageName)
    val pending = if (launch == null) null else PendingIntent.getActivity(
        ctx, 0, launch,
        PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
    )
    return NotificationCompat.Builder(ctx, CHANNEL_ID)
        // A framework status drawable on purpose: it is guaranteed to exist and
        // to be a real status-bar icon. The app's own launcher icon is an
        // adaptive mipmap, which renders as a blank square in the status bar on
        // several Android versions.
        .setSmallIcon(android.R.drawable.stat_sys_download_done)
        .setContentTitle(title)
        .setContentText(body)
        .setStyle(NotificationCompat.BigTextStyle().bigText(body))
        .setOngoing(ongoing)
        .setAutoCancel(!ongoing)
        .setContentIntent(pending)
        .build()
}

/* ── The Activity: receives the share, decides, and gets out of the way ───── */

/**
 * Translucent and content-less, so the user keeps looking at Instagram. It does
 * no network at all — it hands the URL to the service and finishes, which is
 * what makes the share feel instantaneous.
 */
class ShareActivity : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val url = extractUrl(intent?.getStringExtra(Intent.EXTRA_TEXT))

        if (url == null) {
            finish()
            return
        }

        val config = readConfig(this)
        if (config == null) {
            // No key yet: never signed in on this device, or this build's first
            // run. Fall back to Phase A — open the app WITH the share so the
            // existing JS handler saves it. A flash beats a lost link.
            //
            // The component is explicit, so no intent filter has to match; the
            // type must be set because expo-share-intent's activity listener
            // ignores an intent whose `type` is null.
            val launch = packageManager.getLaunchIntentForPackage(packageName)
            if (launch != null) {
                launch.action = Intent.ACTION_SEND
                launch.type = "text/plain"
                launch.putExtra(Intent.EXTRA_TEXT, url)
                launch.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                try { startActivity(launch) } catch (t: Throwable) { }
            }
            finish()
            return
        }

        val work = Intent(this, ShareSaveService::class.java)
            .putExtra(ShareSaveService.EXTRA_URL, url)
            .putExtra(ShareSaveService.EXTRA_API, config.apiUrl)
            .putExtra(ShareSaveService.EXTRA_KEY, config.key)
        try {
            // Legal without restriction because we are the foreground app at
            // this instant — the service promotes itself before that stops
            // being true.
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) startForegroundService(work)
            else startService(work)
        } catch (t: Throwable) {
        }
        finish()
    }
}

/**
 * Instagram and YouTube share a bare URL; other apps share "Look at this <url>"
 * or append a title. Same rule as the JS handler — pluck the first URL rather
 * than refusing anything that is not exactly a link.
 */
internal fun extractUrl(text: String?): String? {
    if (text.isNullOrBlank()) return null
    val m = Regex("https?://\\S+").find(text) ?: return null
    return m.value
}

/* ── The Service: does the save after the user has already walked away ────── */

/**
 * A foreground service, deliberately.
 *
 * The obvious cheaper version — a bare thread started from the Activity — works
 * right up until it doesn't: once the Activity finishes, the process is a
 * cached process and Android may reap it mid-request. On the free Render tier a
 * cold instance takes ~50 s to wake, which is exactly the window where that
 * happens, and the user would get no save AND no notification, having seen
 * nothing at all. This is Android's sanctioned mechanism for short work that
 * outlives the screen, and its "Saving…" notification is honest feedback during
 * that cold start instead of silence.
 */
class ShareSaveService : Service() {

    companion object {
        const val EXTRA_URL = "url"
        const val EXTRA_API = "api"
        const val EXTRA_KEY = "key"
        private const val ONGOING_ID = 4201
        private const val RESULT_ID = 4202
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val url = intent?.getStringExtra(EXTRA_URL)
        val api = intent?.getStringExtra(EXTRA_API)
        val key = intent?.getStringExtra(EXTRA_KEY)

        ensureChannel(this)
        promote()

        if (url == null || api == null || key == null) {
            stopSelf(startId)
            return START_NOT_STICKY
        }

        Thread {
            val where = platformLabel(url)
            val ok = try { postSave(api, key, url) } catch (t: Throwable) { false }
            val title = if (ok) "Saved from $where" else "Couldn't save that $where link"
            val body = if (ok)
                "Your summary is being written — it will be ready in your library."
            else
                "Open SaveHere and paste it to try again."

            // Recorded before it is posted, and recorded even if posting is
            // refused: a user who declined notifications still needs one place
            // where the save is reported.
            appendNote(this, title, body)
            try {
                val mgr = getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
                mgr?.notify(RESULT_ID, buildNotification(this, title, body, false))
            } catch (t: Throwable) {
            }
            stopSelf(startId)
        }.start()

        return START_NOT_STICKY
    }

    private fun promote() {
        val note = buildNotification(this, "Saving to SaveHere", "Sending the link…", true)
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                startForeground(ONGOING_ID, note, ServiceInfo.FOREGROUND_SERVICE_TYPE_DATA_SYNC)
            } else {
                startForeground(ONGOING_ID, note)
            }
        } catch (t: Throwable) {
            // Android 12+ can refuse a foreground start in edge cases. The work
            // still runs on the thread below; it just loses its protection.
        }
    }

    /** True when the card exists server-side. The summary continues on the
     *  backend afterwards, exactly as for a save made in the app. */
    private fun postSave(api: String, key: String, url: String): Boolean {
        var conn: HttpURLConnection? = null
        return try {
            conn = URL("$api/api/reels/share-save").openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json")
            conn.setRequestProperty("X-Share-Key", key)
            conn.connectTimeout = 15000
            // Generous: a cold free-tier Render instance takes ~50 s to wake,
            // and giving up early would report a failure for a save that was
            // about to succeed.
            conn.readTimeout = 90000
            conn.doOutput = true
            conn.outputStream.use { it.write(JSONObject().put("url", url).toString().toByteArray(Charsets.UTF_8)) }
            val code = conn.responseCode
            // 409-style duplicates are still "it's in your library".
            code in 200..299 || code == 409
        } catch (t: Throwable) {
            false
        } finally {
            try { conn?.disconnect() } catch (t: Throwable) { }
        }
    }
}

/** Human name for the notification, from the URL alone — mirrors
 *  `platformLabel` in mobile/services/shareSave.ts. */
internal fun platformLabel(url: String): String {
    val u = url.lowercase()
    return when {
        u.contains("youtube.com") || u.contains("youtu.be") -> "YouTube"
        u.contains("instagram.com") -> "Instagram"
        u.contains("facebook.com") || u.contains("fb.watch") || u.contains("fb.com") -> "Facebook"
        u.contains("tiktok.com") -> "TikTok"
        u.contains("linkedin.com") -> "LinkedIn"
        else -> "the web"
    }
}
